# Regex

## Regular Expression

A regular expression, often called a pattern, is **an expression used to specify a set of strings** required for a particular purpose. 

- A simple way to specify a finite set of strings is to list its elements or members. <br>For example `{Doc1,1,Doc2,2,Doc3,3}`. 
    

`{Doc1,Doc2,Doc3}` can be specified by the pattern `Doc(1|2|3)`. <br>We say that this pattern matches each of the two strings. [Lets check?](https://regex101.com/)   
[Regexr](https://regexr.com/)

> In most formalisms, if there exists at least one regular expression that matches a particular set then there exists an infinite number of other regular expressions that also match it, i.e. **the specification is not unique**.<br>
For example, the string set `{Doc1,Doc2,Doc3}` can also be specified by the pattern `Doc\d`.


In [ ]:
#!pip install regex

In [1]:
import re

## Uses of Regular Expressions

**Some important usages of regular expressions are:**

- Check if an input honors a given pattern; for example, we can check whether a value entered in a HTML formulary is a valid e-mail address
> `Maniteja123@gmail.com`

- Look for a pattern appearance in a piece of text; for example, check if either the word "color" or the word "colour" appears in a document with just **one scan**
> `I like Red color and i am wearing a Red colour shirt`

- Extract specific portions of a text; for example, extract the postal code of an address
> `Mr John Smith. 132, My Street, Kingston, New York 12401.`

- Replace portions of text; for example, change the appearance of "color" with "colour"
> `I like Red colour and i am wearing a Red colour shirt`

- Split a larger text into smaller pieces, for example, splitting a text by any appearance of the dot, comma, or newline characters
> `myself person1,you are person2`

# Meta Characters

- All meta characters. `^ $ * + ? { } \ | ( ) `

  1. `.` any character (except new line character)
  2. `^` startswith `^word`
  3. `$` endswith `word$`
  4. `*` zero or more occurrences
  5. `+` one or more occurrences
  6. `{}` exactly specified no of occurrences "M{2}"
  7. `[]` A set of characters "[a-c]"
  8. `\` Signals a special sequence (can also be used to escape special characters) "\d"
  9. `|` Either or "apple|iphone"
  10. `()` Capture and group

In [9]:
a = "hat hit hot hut h6t h%t h t"
print(re.findall(r"h.t",a))
print(re.findall(r"h\wt",a))  # a-z-0-9-_
print(re.findall(r"h\dt",a))
print(re.findall(r"h\Wt",a)) #^a-z-0-9-_

['hat', 'hit', 'hot', 'hut', 'h6t', 'h%t', 'h t']
['hat', 'hit', 'hot', 'hut', 'h6t']
['h6t']
['h%t', 'h t']


In [16]:
b= "hello world"
print(re.findall(r"^z",b))
print(re.findall(r"^h",b))
print(re.findall(r"^@",b))

# \A
print(re.findall(r"\A@",b))


[]
['h']
[]
[]


In [38]:
b= """hello world 
     \nhello world 
     \nhello world"""

print(re.findall(r"\Ahello",b,re.MULTILINE))
print(re.findall(r"^hello",b,re.MULTILINE))
print()

['hello']
['hello', 'hello', 'hello']
hello world 
hello world 
hello world


In [30]:
c = "hello world"

print(re.findall(r"world$",b))


['world']


In [44]:
b= """hello world
      \nhello world
      \nhello world"""

print(re.findall(r"world\Z",b,re.MULTILINE))
print(re.findall(r"world$",b,re.MULTILINE))
print(b)

['world']
['world', 'world', 'world']
hello world
      
hello world
      
hello world


In [52]:
d = "a ab abb abab ababab"   
print(re.findall(r"ab*",d))
print(re.findall(r"b*",d))
print(re.findall(r"(ab)*",d)) 

['a', 'ab', 'abb', 'ab', 'ab', 'ab', 'ab', 'ab']
['', '', '', 'b', '', '', 'bb', '', '', 'b', '', 'b', '', '', 'b', '', 'b', '', 'b', '']
['', '', 'ab', '', 'ab', '', '', 'ab', '', 'ab', '']


In [60]:
e = "a aab abb abbb"
print(re.findall(r"ab+",e))
print(re.findall(r"a+b+",e))
print(re.findall(r"a+b",e))
print(re.findall(r"(a+)",e))

['ab', 'abb', 'abbb']
['aab', 'abb', 'abbb']
['aab', 'ab', 'ab']
['a', 'aa', 'a', 'a']


In [63]:
f = "MM M MMM"

print(re.findall(r"M{2,}",f)) # two or more
print(re.findall(r"M{2}",f))

['MM', 'MMM']
['MM', 'MM']


In [66]:
print(re.findall(r"\w+.{2}\w+","pattern"))
print(re.findall(r"\w+.{2}\w+","hello"))

['pattern']
['hello']


In [69]:
g = "abcdef"
print(re.findall(r"[a-c]",g))
print(re.findall(r"[a-c]+",g))
print(re.findall(r"^[a-c]+",g))

['a', 'b', 'c']
['abc']
['abc']


In [84]:
g = "apple \nball \ncat \ndog \negg \nfish"
print(re.findall(r"^[a-c]\w+",g,re.M))
print(g)

['apple', 'ball', 'cat']
apple 
ball 
cat 
dog 
egg 
fish


In [91]:
h = "abc123.abc123"
print(re.findall(r"\d+",h))
print(re.findall(r"\w+",h))
print(re.findall(r"\..+",h))

['123', '123']
['abc123', 'abc123']
['.abc123']


In [93]:
k = "apple iphone samsung Apple color colour"
print(re.findall(r"apple|Apple",k))
print(re.findall(r"color|colour",k))

['apple', 'Apple']
['color', 'colour']


In [112]:
l =  "ab abab python ababab"
print(re.findall(r"(ab).+",l))
print(re.findall(r"abab\s(.+)\s",l))

['ab']
['python']


# Special Sequences
- A special sequence is a \ followed by one of the characters in the list below, and has a special       meaning:

  1. `\d` : Matches any decimal digit; this is equivalent to the class [0-9].
  2. `\D` : Matches any non-digit character; this is equivalent to the class [^0-9].
  3. `\s` : Matches any whitespace character, next line character(\n) or tab(\t);
  4. `\S` : Matches any non-whitespace character;
  5. `\w` : Matches any alphanumeric (word) character; this is equivalent to the class [a-zA-Z0-9_].
  6. `\W` : Matches any non-alphanumeric character; this is equivalent to the class [^a-zA-Z0-9_].

In [114]:
d1 = "ID123"
print(re.findall(r"\d+",d1))

['123']


In [129]:
e1 = """ID123#*$%"""
print(re.findall(r"\D+",e1))

['ID', '#*$%']


In [131]:
f1 = "a b\tc\nd"
print(f1)
print(re.findall(r"\s+",f1))

a b	c
d
[' ', '\t', '\n']


In [137]:
g1 = "a b\tc\nd @#$"
print(re.findall(r"\S+",g1))

['a', 'b', 'c', 'd', '@#$']


In [139]:
h1 = "user_123@jio"
print(re.findall(r"\w+",h1))

['user_123', 'jio']


In [140]:
h1 = "user_123"
print(re.findall(r"\W+",h1))

[]


In [143]:
i1 = "user@123 %& *("
print(re.findall(r"\W+",i1))

['@', ' %& *(']


#### Some examples for set 
 1. `[arn]` Returns a match where one of the specified characters (a, r, or n) are present
 2. `[a-n]` Returns a match for any lower case character, alphabetically between a and n
 3. `[^arn]` Returns a match for any character EXCEPT a, r, and n
 4. `[0123]` Returns a match where any of the specified digits (0, 1, 2, or 3) are present
 5. `[0-9]` Returns a match for any digit between 0 and 9
 6. `0-5` Returns a match for any two-digit numbers from 00 and 59
 7. `[a-zA-Z]` Returns a match for any character alphabetically between a and z, lower case OR upper      case

In [144]:
colleges="""college : IIT Madras - Indian Institute of Technology
4.6(202)
Fees: ₹ 10.00 Lakh
Salary : ₹ 16.00 Lakh
Not Ranked
Times ' 22
3
The Week ' 21
1
Outlook ' 20
Admissions
Courses & Fees
Placements 0-5
college : IIT Madras - Indian Institute of Technology
4.1(202)
Fees: ₹ 10.00 Lakh
Salary : ₹ 16.00 Lakh
Not Ranked
Times ' 22
3
The Week ' 21
1
Outlook ' 20
Admissions
Courses & Fees
Placements 0-5
college : IIT Madras - Indian Institute of Technology
4.4(202)
Fees: ₹ 10.00 Lakh
Salary : ₹ 16.00 Lakh
Not Ranked
Times ' 22
3
The Week ' 21
1
Outlook ' 20
Admissions
Courses & Fees
Placements 0-5"""

In [147]:
# extract college name
print(re.findall(r"college : (.+)",colleges))
print(re.findall(r"college\s\W+\s(.+)",colleges))

['IIT Madras - Indian Institute of Technology', 'IIT Madras - Indian Institute of Technology', 'IIT Madras - Indian Institute of Technology']
['IIT Madras - Indian Institute of Technology', 'IIT Madras - Indian Institute of Technology', 'IIT Madras - Indian Institute of Technology']


In [154]:
# extract the fees
print(re.findall(r"Fees: (.+)",colleges))
print(re.findall(r"Fees: \W+(.+) Lakh",colleges))
print(re.findall(r"Fees: \W+(\d+.\d+)",colleges))

['₹ 10.00 Lakh', '₹ 10.00 Lakh', '₹ 10.00 Lakh']
['10.00', '10.00', '10.00']
['10.00', '10.00', '10.00']


In [158]:
# extact the Placements
print(re.findall(r"Placements\s(.+)",colleges))
print(re.findall(r"Placements\s(\d+\D+\d+)",colleges))
print(re.findall(r"Placements\s(\d+\W+\d+)",colleges))
print(re.findall(r"Placements\s(\d+.\d+)",colleges))

['0-5', '0-5', '0-5']
['0-5', '0-5', '0-5']
['0-5', '0-5', '0-5']
['0-5', '0-5', '0-5']


In [160]:
print(re.findall(r"(\d+.\d+)\(",colleges))

['4.6', '4.1', '4.4']


In [6]:
emails="""python@gmail.com
python_sql@gmail.com
python.sql@ gmail.com
python1234@gmail.com
python@outlook.com
python_sql@outlook.com
python.sql@outlook.com
python1234@outlook.com
python@yahoo.org
python_sql@yahoo.com
python.sql@yahoo.in
python1234@yahoo.com
"""

In [161]:
#Extracting username of all emails
# Extract usernames which has number in it
#Extracting domain of all emails
#Extracting com,org,in of all emails
#Extracting emails of outlook 
#Extracting all emails 

In [7]:
#Extracting username of all emails
import re
print(re.findall(r"(\w+)@",emails))

['python', 'python_sql', 'sql', 'python1234', 'python', 'python_sql', 'sql', 'python1234', 'python', 'python_sql', 'sql', 'python1234']


In [10]:
# Extract usernames which has number in it
print(re.findall(r"(.+\d+)@",emails))

['python1234', 'python1234', 'python1234']


In [11]:
#Extracting domain of all emails
print(re.findall(r"@(\w+).",emails))

['gmail', 'gmail', 'gmail', 'outlook', 'outlook', 'outlook', 'outlook', 'yahoo', 'yahoo', 'yahoo', 'yahoo']


In [12]:
#Extracting com,org,in of all emails
print(re.findall(r"@.+\.(\w+)",emails))

['com', 'com', 'com', 'com', 'com', 'com', 'com', 'com', 'org', 'com', 'in', 'com']


In [13]:
print(re.findall(r".+outlook.+",emails))

['python@outlook.com', 'python_sql@outlook.com', 'python.sql@outlook.com', 'python1234@outlook.com']


In [4]:
phones="""SAMSUNG Galaxy F23 5G (Forest Green, 128 GB)
SAMSUNG Galaxy F04 (Jade Purple, 64 GB)
POCO M3 Pro 5G (Yellow, 128 GB)
MOTOROLA e40 (Carbon Gray, 64 GB)
APPLE iPhone 13 (Blue, 128 GB)
APPLE iPhone 14 (Starlight, 128 GB)
APPLE iPhone 14 (Blue, 128 GB)
MOTOROLA G62 5G (Midnight Gray, 128 GB)
REDMI 10 (Pacific Blue, 64 GB)
REDMI 10 (Caribbean Green, 64 GB)
REDMI Note 11 SE (Cosmic White, 64 GB)
MOTOROLA G32 (Mineral Gray, 64 GB)
MOTOROLA G62 5G (Frosted Blue, 128 GB)
POCO C31 (Royal Blue, 64 GB)
MOTOROLA e40 (Pink Clay, 64 GB)
REDMI 10 (Midnight Black, 64 GB)
SAMSUNG Galaxy F23 5G (Copper Blush, 128 GB)
REDMI Note 12 Pro+ 5G (Obsidian Black, 256 GB)
SAMSUNG Galaxy F04 (Opal Green, 64 GB)
MOTOROLA G32 (Satin Silver, 64 GB)
POCO M4 Pro (Cool Blue, 64 GB)
REDMI 9i Sport (Coral Green, 64 GB)
SAMSUNG Galaxy F23 5G (Forest Green, 128 GB)
POCO M4 Pro (Cool Blue, 128 GB)"""

In [ ]:
#Extracting mobile Brands
#Extracting the mobile version
#Extracting storage of mobile phones
#Extracting color of mobile phones

# Functions

`findall()` : Returns a list containing all matches    
>  **syntax** : re.findall (pattern, target string)

`search` : Searches for the first match anywhere in the string.   
`match` : Matches only at the beginning of the string.  
`split`      : Splits string based on a regex pattern.    
`sub`        : Replaces matched patterns.   
`subn`       : Same as `sub()`, but also returns count of replacements..

In [1]:
import re

In [2]:
a = "ID 123 and 45"

In [3]:
re.findall(r"\d+",a)

['123', '45']

In [18]:
print(re.search(r"\d+",a))
print(re.search(r"\d+",a).group())
print(re.search(r"\s(\d+)\s\w+\s(\d+)",a).groups())
print(re.search(r"\s(\d+)\s\w+\s(\d+)",a).group())
print(re.search(r"\s(\d+)\s\w+\s(\d+)",a).groups()[0])
print(re.search(r"\s(\d+)\s\w+\s(\d+)",a).span())

<re.Match object; span=(3, 6), match='123'>
123
('123', '45')
 123 and 45
123
(2, 13)


In [38]:
a = "ID 123 and 45"
print(re.match(r"\d+",a))
print(re.match(r"\w+\s+(\d+)",a).group())
print(re.match(r"(\w+)\s+(\d+)",a).groups(1))
print(re.match(r"(\w+)\s+(\d+)",a).group(2))

None
ID 123
('ID', '123')
123


In [22]:
str1 = "123 and 45"
re.match(r"\d+",str1)

<re.Match object; span=(0, 3), match='123'>

In [52]:
str2 = "ID   123 and 45"
str3 = "ID%123 and$45"

print(re.split(r"\s+",str2))
print(re.split(r"\s+",str3))
print(re.split(r"\W+",str3))
print(re.split(r"\w+",str3))

['ID', '123', 'and', '45']
['ID%123', 'and$45']
['ID', '123', 'and', '45']
['', '%', ' ', '$', '']


In [58]:
str3 = "ID%123 and$45"
re.sub(r"\W+","*",str3) # ("regex","parameter",string)

'ID*123*and*45'

In [59]:
str3 = "ID%123 and$45"
re.subn(r"\W+","*",str3)

('ID*123*and*45', 3)

# Flages 

Regex flags modify how a pattern is interpreted.   
They are passed as an optional argument to regex functions.

> **syntax:** re.findall(pattern, string, flags)

- `re.I or re.IGNORECASE` : Makes pattern case-insensitive.
- `re.M or re.MULTILINE`  : Changes behavior of `^` and `$` to match each line, not just entire string.
- `re.S or re.DOTALL`     : Makes `.` match newline `(\n)` also.
- `re.X or re.VERBOSE`    : Allows whitespace and comments in regex for readability.
- `re.A or re.ASCII`      : Makes `\w, \d, \s, \b` match ASCII only.


In [62]:
s1 = "Python PYTHON python"

print(re.findall(r"python",s1,re.IGNORECASE))
print(re.findall(r"python",s1))

['Python', 'PYTHON', 'python']
['python']


In [67]:
text = """hello
world
hello"""

print(re.findall(r"^w",text,re.M))

['w']


In [73]:
text = "hello\nworld"
print(re.findall(r"\w+",text))
print(re.findall(r".+",text))
print(re.findall(r"\w+\W+\w+",text))
print(re.findall(r".+",text,re.S))
print(re.findall(r"\w+",text,re.S)) #a-z-0-9-_
print(re.findall(r".+",text,re.DOTALL))

['hello', 'world']
['hello', 'world']
['hello\nworld']
['hello\nworld']
['hello', 'world']
['hello\nworld']


In [82]:
pattern = r"""
\d{3}      # area code
-          # dash
\d{3}      # first 3 digits
-          # dash
\d{4}      # last 4 digits
"""
s2 = "123-456-7890"

print(re.findall("\d+-\d+-\d+",s2))
print(re.findall("\d{3}-\d{3}-\d{4}",s2))
print(re.findall(".+",s2))
print(re.findall(pattern,s2,re.VERBOSE))
print(re.findall(pattern,s2))

['123-456-7890']
['123-456-7890']
['123-456-7890']
['123-456-7890']
[]


In [85]:
s3 = "naïve café %^"
re.findall(r"\w+",s3,re.A)

['na', 've', 'caf']

### TASK

In [90]:
# Extract all numbers from the text.
text = "Order123 shipped on 25-08-2025 with ID4567"

print(re.findall("\d+",text))

# Extracting the date
print(re.findall("\d{2}[-]\d{2}[-]\d{4}",text))
print(re.findall("\d{2}-\d{2}-\d{4}",text))
print(re.findall("\d+\W+\d+\W+\d+",text))

['123', '25', '08', '2025', '4567']
['25-08-2025']
['25-08-2025']
['25-08-2025']


In [94]:
# Extract only 10-digit phone numbers.
phones = "9876543210, 12345, 9123456789, 999"
print(re.findall("\d{10}",phones))

['9876543210', '9123456789']


In [103]:
# Match all occurrences of "python" ignoring case.
text = "Python python PYTHON PyThOn"

print(re.findall("python",text,re.IGNORECASE))
print(re.findall(r".+",text,re.IGNORECASE))
print(re.findall(r"python", text, re.IGNORECASE))
print(re.findall('python',text,re.I))

['Python', 'python', 'PYTHON', 'PyThOn']
['Python python PYTHON PyThOn']
['Python', 'python', 'PYTHON', 'PyThOn']
['Python', 'python', 'PYTHON', 'PyThOn']


In [114]:
# Match only standalone word "cat".
# Match "cat" inside words.
sentence = "cat scAtter copycat cat"

print(re.findall(r"\bcat",sentence,re.I))
print(re.findall(r"\bcat\b",sentence,re.I))
print(re.findall(r"cat\b",sentence,re.I))
print(re.findall(r"\Bcat\B",sentence,re.I))


['cat', 'cat']
['cat', 'cat']
['cat', 'cat', 'cat']
['cAt']


In [117]:
# Replace all digits with *.
data = "User123 has balance 4500"
print(re.sub(r'\d','*',data))

User*** has balance ****


In [ ]:
# Extract all lines starting with "error".
text = """error: file missing
info: process started
error: permission denied"""



In [ ]:
# Match everything from "start" to "end".
text = "start\nmiddle\nend"


In [ ]:
# Extract all email urseid's
emails = "test@gmail.com admin@yahoo.in user@company.org"


In [ ]:
#Check which IDs contain only digits and are exactly 5 characters long.
ids = ["12345", "1234a", "67890"]


In [ ]:
# Write a readable regex to match phone numbers like:
num = "123-456-7890"

In [ ]:
# Extract the error part
# Extract the paths as well
log = "ERROR 404 at /home | INFO 200 at /login | ERROR 500 at /api"
